src/libcore/iter/iterator.rs

```diff
@@ -1337,6 +1335,71 @@ pub trait Iterator {
         (left, right)
     }
 
+    /// An iterator method that applies a function as long as it returns
+    /// successfully, producing a single, final value.
+    ///
+    /// `fold()` takes two arguments: an initial value, and a closure with two
+    /// arguments: an 'accumulator', and an element. The closure either returns
+    /// successfully, with the value that the accumulator should have for the
+    /// next iteration, or it returns failure, with an error value that is
+    /// propagated back to the caller immediately (short-circuiting).
+    ///
+    /// The initial value is the value the accumulator will have on the first
+    /// call.  If applying the closure succeeded against every element of the
+    /// iterator, `try_fold()` returns the final accumulator as success.
+    ///
+    /// Folding is useful whenever you have a collection of something, and want
+    /// to produce a single value from it.
+    ///
+    /// # Note to Implementors
+    ///
+    /// If possible, override this method with an implementation using
+    /// internal iteration.  Most of the other methods have their default
+    /// implementation in terms of this one.
```

Comment:
Hm, any implementation is by definition using internal iteration? Maybe we can say they should in turn use `try_fold` on the parts the iterator are composed of, if possible. And some kind of smooth reminder that the implementation must keep all internal state up to date, no matter if it finishes with Ok or Err.

Reply:
I took another stab at this section; hopefully it's on the right track.